src/components/ConfigEditor/ConfigEditor.jsx

```diff
@@ -0,0 +1,178 @@
+import React, { useState, useEffect } from "react";
+import useSWR from "swr";
+import { getHintOptions } from "vitessce";
+
+import { StudyIdInput } from "../StudyIdInput";
+import { baseJson, exampleConfig } from "../../utils/config-examples.js";
+import { ErrorDiv } from "../ErrorDiv";
+
+import styles from "./ConfigEditor.module.css";
+
+import {
+	ERROR_MESSAGES,
+	EXAMPLE_URL,
+	LINK_ID_ENDPOINT_URL,
+	LINK_ID_LENGTH,
+	LINK_ID_KEY,
+} from "../../utils/constants.js";
+import {
+	validateConfig,
+	sanitizeURLs,
+	studyIdFetcher,
+	updateConfigWithExampleURL,
+} from "../../utils/utility-functions.js";
+
+export const ConfigEditor = ({
+	pendingJson,
+	setPendingJson,
+	error,
+	setError,
+	setLoading,
+	setUrl,
+	setLinkIdInput,
+}) => {
+	const [datasetUrls, setDatasetUrls] = useState("");
+	const [generateConfigError, setGenerateConfigError] = useState(null);
+	const [studyId, setStudyId] = useState(null);
+	const [linkId, setLinkId] = useState(null);
+	const [linkIdUrl, setLinkIdUrl] = useState(null);
+	const [showReset, setShowReset] = useState(null);
+
+	function handleGetLinkId(studyId) {
+		setStudyId(studyId);
+	}
+
+	useSWR(linkIdUrl, studyIdFetcher, {
+		onError: (err) => {
+			setError(err.message);
+			setLoading(false);
+		},
+		onSuccess: (data) => {
+			if (data && data[LINK_ID_KEY]) {
+				setLinkId(data[LINK_ID_KEY]);
+			}
+			setLoading(false);
+		},
+		shouldRetryOnError: false,
+	});
+
+	useEffect(() => {
+		if (studyId) {
+			const constructedUrl = `${LINK_ID_ENDPOINT_URL}${studyId}`;
+			setLinkIdUrl(constructedUrl);
+		}
+	}, [studyId]);
+
+	function handleInputError(errMessage) {
+		setError(errMessage);
+	}
+
+	async function handleLaunch() {
+		if (datasetUrls === "") {
+			setError(ERROR_MESSAGES.NO_DATASET_URL_ERROR);
+			return;
+		}
+		if (error || generateConfigError) return;
+
+		if (linkId && linkId !== "" && String(linkId).length === LINK_ID_LENGTH) {
+			let nextUrl;
+			const nextConfig = pendingJson;
+			nextUrl = `data:,${encodeURIComponent(nextConfig)}`;
+			const [valid, failureReason] = validateConfig(nextConfig);
+			if (!valid) {
+				setError(failureReason);
+				return;
+			}
```

Comment:
Since the `nextUrl` gets dropped if validation fails anyway, we can reorganize slightly to get rid of the `let`:

```suggestion
			const nextConfig = pendingJson;
			const [valid, failureReason] = validateConfig(nextConfig);
			if (!valid) {
				setError(failureReason);
				return;
			}
			const nextUrl = `data:,${encodeURIComponent(nextConfig)}`
```